Fix handling of deactivated custom emoji in user status.

YashRE42 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
I've opened this wip PR to get feedback on some changes:
I would be glad to simply implement the change suggested in the comment added by the first commit, rather than leaving the comment.
The third commit needs work, but I'm not certain about which direction we want the codebase to move towards.

Pasting the contents of the last commit message:

Previously, if a user had set a realm emoji as their status emoji and
someone would then delete the realm emoji, the app would fail to
initialize, because of the error we throw from `emoji.js`.

The current commit fixes this by just displaying the deactivated
emoji, similar to how we do when realm_emoji used as reactions are
deleted.

However, this commit has to resort to somewhat hacky additions to
solve the bug, which may also just be a symptom of other refactors
that are needed.

In addition, this also suffers from the flaw that if a new emoji is
added with the same name as the previously deleted emoji, then the
user status will show the new emoji. This could lead to shenanigans
and is why this is labeled as do_not_merge.

Fixes: #20274.


**Testing plan:** <!-- How have you tested? -->
added a case to user_status node test.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
![191121deactivated-status-emoji](https://user-images.githubusercontent.com/33805964/142645151-dbd63ce5-af65-4c33-bf5a-6990db663cbe.gif)


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

timabbott commented:
> In addition, this also suffers from the flaw that if a new emoji is
added with the same name as the previously deleted emoji, then the
user status will show the new emoji. This could lead to shenanigans
and is why this is labeled as do_not_merge.

I think this is what our design should intend; the logical purpose of deleting a custom emoji and re-uploading is to improve the image being used for the concept in your organization.

timabbott commented on static/shared/js/emoji.js:
```diff
@@ -170,7 +171,15 @@ export function update_emojis(realm_emojis) {
+    // as its ID. All other emoji use numeric IDs. We should avoid
+    // creating a special case like this as it's quite liable to cause
+    // bugs. We might be interested in giving it the ID of 0 (if that's
+    // safe to do?).
```
I think our plan for the Zulip emoji is to convert it into a RealmEmoji at some point, but it's a big refactor that's probably not worth it for what we might game.

timabbott commented on static/shared/js/emoji.js:
```diff
@@ -208,12 +215,24 @@ export function get_emoji_details_by_name(emoji_name) {
-        const codepoint = get_emoji_codepoint(emoji_name);
-        if (codepoint === undefined) {
-            throw new Error("Bad emoji name: " + emoji_name);
+        // deactivation is implied by reaching this line
```
This comment isn't quite right -- there is the further else clause below.

timabbott commented on static/shared/js/emoji.js:
```diff
@@ -168,6 +168,13 @@ export function update_emojis(realm_emojis) {
+            // shift everywhere else from using emoji_names to emoji_codes,
+            // which is probably better in the long term, but more complex
+            // as it means we'd need to go up all the way to the data-attributes
+            deactivated_emoji_name_to_code.set(data.name, data.id);
```
Do we need this data structure, or could we just do the lookup in `all_realm_emojis`?

timabbott commented on static/shared/js/emoji.js:
```diff
@@ -208,12 +215,24 @@ export function get_emoji_details_by_name(emoji_name) {
+                emoji_info.still_url = deactivated_realm_emoji.still_url;
+            }
+        } else {
+            const codepoint = get_emoji_codepoint(emoji_name);
```
I think we want this check, for unicode emoji, to be before the deactivated emoji check.

timabbott commented:
I'm not sure the strategy taken in this PR is the right one.  Read `insert_new_reaction` for an eample.  I think basically the `get_emoji_details_by_name` function isn't the right thing to call when we already have an object from the server that specifically has the name/field/reaction_type values.  Here's what we're doing in the status code:

```
    user_status_emoji_info.set(opts.user_id, {                                                           
        emoji_name: opts.emoji_name,                                                                     
        emoji_code: opts.emoji_code,                                                                     
        reaction_type: opts.reaction_type,                                                               
        emoji_alt_code: user_settings.emojiset === "text",                                               
        ...emoji.get_emoji_details_by_name(opts.emoji_name),                                             
    });                                                                                                  
```

In contrast, here's what the reaction code has us pass into the template:

```
<div class="{{this.class}}" aria-label="{{this.label}}" data-reaction-id="{{this.local_id}}">            
    {{#if this.emoji_alt_code}}                                                                          
        <div class="emoji_alt_code">&nbsp:{{this.emoji_name}}:</div>                                     
    {{else}}                                                                                             
        {{#if this.is_realm_emoji}}                                                                      
        <img src="{{this.url}}" class="emoji" />                                                         
        {{else}}                                                                                         
        <div class="emoji emoji-{{this.emoji_code}}"></div>                                              
        {{/if}}                                                                                          
    {{/if}}                                                                                              
    <div class="message_reaction_count">{{this.count}}</div>                                             
</div>                                                                                                   
```

There's no lookup at all required for unicode emoji; so it just does something like this:
```
    const context = {                                                                                    
        message_id,                                                                                      
        emoji_name,                                                                                      
        emoji_code,                                                                                      
    };                                                                                                   
                                                                                                         
    const new_label = generate_title(emoji_name, user_list);                                             
                                                                                                         
    if (opts.reaction_type !== "unicode_emoji") {                                                        
        context.is_realm_emoji = true;                                                                   
        const emoji_info = emoji.all_realm_emojis.get(emoji_code);                                       
        if (!emoji_info) {                                                                               
            blueslip.error(`Cannot find/insert realm emoji for code '${emoji_code}'.`);                  
            return;                                                                                      
        }                                                                                                
        context.url = emoji_info.emoji_url;                                                              
    }                                                                                                    
                                                                                                         
    context.count = 1;                                                                                   
    context.label = new_label;                                                                           
    context.local_id = get_local_reaction_id(opts);                                                      
    context.emoji_alt_code = user_settings.emojiset === "text";                                          
                                                                                                         
    ...
```

It'd probably be useful to extract a `get_emoji_details_for_rendering()` function that takes in an object with `reaction_type, emoji_code, emoji_name` and lets us avoid adding a duplicate of that logic for displaying an emoji given those 3 properties.  But basically, the status emoji code path should be calling a function with that signature, not something that just takes a name.

YashRE42 commented:
I've updated this, based on the given feedback.

While working on this I noticed that this feature does not work correctly with the `plain text` emoji display setting.
1. It does not correctly update existing emoji to plain text.
1. The plain text emoji in the set status UI overflow onto the status text.
1. You cannot save from the set status form if you've selected an emoji from the emoji picker.
1. It does not live update a plain text emoji to an emoji theme (it does on refresh).

The only case that seems to work properly is if you have `plain text` set and use one of the 5 predefined emoji.

![201121status-emoji-as-text-bug](https://user-images.githubusercontent.com/33805964/142723502-2e6a3b24-c724-4da8-8d25-c28a8d5c230c.gif)
![201121status-emoji-as-text-bug-2](https://user-images.githubusercontent.com/33805964/142723508-d502e22a-5e44-4a69-9e85-42d054526611.gif)

timabbott commented:
@YashRE42 this is great, but I pushed a commit we can squash that I think would provide a cleaner implementation logically, in terms of doing lookups based on `reaction_type` and then `emoji_code`.  I think you can then add a few commits to make the reactions code call this new function rather than duplicating it.  

(BTW, that deduplication will help us fix the bug that reactions don't support `still_url`, which I think @Riken-Shah might be working on?)

YashRE42 commented:
Apologies for not understanding the intent earlier!
I changed the function slightly (see bellow), squashed and replaced the two instances of duplication in reactions.

```js

export function get_emoji_details_for_rendering(opts) {
    if (!opts.emoji_name || !opts.emoji_code || !opts.reaction_type) {
        throw new Error("Invalid params.");
    }

    if (opts.reaction_type !== "unicode_emoji") {
        const realm_emoji = all_realm_emojis.get(opts.emoji_code);
        if (!realm_emoji) {
            throw new Error(`Cannot find realm emoji for code '${opts.emoji_code}'.`);
        }
        return {
            url: realm_emoji.emoji_url,
            still_url: realm_emoji.still_url,
            emoji_name: opts.emoji_name,
            emoji_code: opts.emoji_code,
            reaction_type: opts.reaction_type,
        };
    }
    // else
    return {
        emoji_name: opts.emoji_name,
        emoji_code: opts.emoji_code,
        reaction_type: opts.reaction_type,
    };
}
```

timabbott commented:
This is great, merged, thanks @YashRE42!

I think now https://github.com/zulip/zulip/issues/13434 should be relatively quick, since we've now plumbed through the `still_url` parameter to the reactions codebase.
